sklearn/metrics/classification.py

```diff
@@ -1596,37 +1603,63 @@ def log_loss(y_true, y_pred, eps=1e-15, normalize=True, sample_weight=None):
     -----
     The logarithm used is the natural logarithm (base-e).
     """
+    y_pred = check_array(y_pred, ensure_2d=False)
+    check_consistent_length(y_pred, y_true)
+
     lb = LabelBinarizer()
-    T = lb.fit_transform(y_true)
-    if T.shape[1] == 1:
-        T = np.append(1 - T, T, axis=1)
 
-    y_pred = check_array(y_pred, ensure_2d=False)
-    # Clipping
-    Y = np.clip(y_pred, eps, 1 - eps)
+    if labels is not None:
+        lb.fit(labels)
+    else:
+        lb.fit(y_true)
 
-    # This happens in cases when elements in y_pred have type "str".
-    if not isinstance(Y, np.ndarray):
+    if len(lb.classes_) == 1:
+        if labels is None:
+            raise ValueError('y_true has only one label. Please provide '
+                             'the true labels explicitly through the labels '
+                             'argument.')
+        else:
+            raise ValueError('labels should have more than one unique label.')
+
+    transformed_labels = lb.transform(y_true)
+
+    if transformed_labels.shape[1] == 1:
+        transformed_labels = np.append(1 - transformed_labels,
+                                       transformed_labels, axis=1)
+    try:
+        y_pred = y_pred.astype(float)
+    except ValueError:
+        raise ValueError("Unable to automatically cast y_pred to "
+                         "float. y_pred should be an array of floats.")
+    # sanity check
+    if y_pred.dtype != float:
```

Comment:
So the correct way to do this is to pass a `dtype` argument to `check_array` that raises an error if it is unable to be cast to the provided dtype. But if I pass `dtype=np.float32`, it fails this test (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/tests/test_classification.py#L1392) because `check_array` raises an error with a `MockDataFrame` and `dtype=np.float32`.

For now, I would suggest to either just remove these float checks (note that nothing useful was checked previously and np.clip should take care of string dtypes) or figure out what is going on with the MockDataframe and fix that (which is beyond the scope of this PR)
Wdyt?